### PR TITLE
feat(protocol): implement request and response headers

### LINKED ISSUE #20: Create RequestHeaders and ResponseHeaders

### PR DESCRIPTION
**Pull request by kordiangrabowski**
closes #20

### CODE REVIEW

File: src/protocol/responses/headers/response-header-v0.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+ * Response Header v0 => correlation_id
+ *   correlation_id => INT32
+ *
+ * @see https://kafka.apache.org/24/protocol.html#protocol_messages
+ */
+export class ResponseHeaderV0 implements ResponseHeader {
+  constructor(public readonly correlationId: Int32) {}
+
+  public static deserialize(buffer: ReadBuffer): ResponseHeaderV0 {
+    return new ResponseHeaderV0(Int32.deserialize(buffer));
```
Comment (Kalin8900) [inline src/protocol/responses/headers/response-header-v0.ts:15]: See: Other inline discussion

File: src/protocol/responses/headers/response-header-v1.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+ * @see https://kafka.apache.org/24/protocol.html#protocol_messages
+ */
+export class ResponseHeaderV1 implements ResponseHeader {
+  constructor(
+    public readonly correlationId: Int32,
+    public readonly tagBuffer: CompactArray<TaggedField>
+  ) {}
+
+  public static deserialize(buffer: ReadBuffer): ResponseHeaderV1 {
+    return new ResponseHeaderV1(Int32.deserialize(buffer), CompactArray.deserialize(buffer, TaggedField.deserialize));
```
Comment (Kalin8900) [inline src/protocol/responses/headers/response-header-v1.ts:19]: See: Other inline discussion

**Review by Kalin8900 — APPROVED** [review]

File: src/protocol/commons/tag-section.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+ * @see https://cwiki.apache.org/confluence/display/KAFKA/KIP-482%3A+The+Kafka+Protocol+should+Support+Optional+Tagged+Fields
+ */
+export class TagSection implements Serializable {
+  constructor(public readonly fields: TaggedField[] = []) {
+    //check the tags ordering
+    for (let i = 0; i < fields.length - 1; i++) {
+      if (fields[i]!.tag.value > fields[i + 1]!.tag.value) {
+        throw new InvalidOrOutOfOrderTagError(`Invalid or out-of-order tag ${fields[i + 1]!.tag.value}`);
+      }
+    }
```
Comment (Kalin8900) [inline src/protocol/commons/tag-section.ts:29]: Do we need to keep the ordering for the whole time? If so we should make the fields private and add methods to mutate it which will keep the order. If not we should move the order check to deserialize and serialize methods.

File: src/protocol/commons/tag-section.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    for (let i = 0; i < numberOfTaggedFields; i++) {
+      taggedFields.push(TaggedField.deserialize(buffer));
+    }
+
+    return new TagSection(taggedFields);
+  }
+
+  public serialize(buffer: WriteBuffer): void {
+    new UVarInt(this.fields.length).serialize(buffer);
+    this.fields.forEach((value) => value.serialize(buffer));
```
Comment (Kalin8900) [inline src/protocol/commons/tag-section.ts:44]: Order might not be correct here see [this comment](https://github.com/trisquareeu/kafkit/pull/33#discussion_r1276838716)

**Review by Kalin8900 — COMMENTED** [review]
I have only one point which we should discuss, but it's well done!

File: src/protocol/commons/tag-section.ts (hunk 3 above)
Comment (kordiangrabowski) [reply src/protocol/commons/tag-section.ts:29]: At this moment, I see no reason for mutating once created object inner fields, therefore I will ensure the field is not exposed for external changes. Good catch, thank you!

**Review by Kalin8900 — APPROVED** [review]

**Pull request merged**